### PR TITLE
pim-3792 : CatalogBundle PHPUnit to phpspec

### PR DESCRIPTION
**Pull request by jmleroux**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | n
| BC breaks?           | n 
| CI currently passes? | y
| Tests pass?          | y
| Scenarios pass?      | y
| Checkstyle issues?*  | n
| PMD issues?**        | y
| Changelog updated?   | n
| Fixed tickets        | pim-3792
| DB schema updated?   | n
| Migration script?    | n
| Doc PR               | n

### CODE REVIEW

File: spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterAttributeConstraintGuessersPassSpec.php [hunk 1]
```diff
@@ -8,27 +8,58 @@
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Definition;
 
-class RegisterAttributeConstraintGuessersPassSpec extends ObjectBehavior {
+class RegisterAttributeConstraintGuessersPassSpec extends ObjectBehavior
+{
+    public function it_is_a_compiler_ass()
```
Comment (jjanvier) [inline spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterAttributeConstraintGuessersPassSpec.php]: don't like this kind of ass ^^

Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterAttributeConstraintGuessersPassSpec.php]: :+1:

File: src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php [hunk 2]
```diff
@@ -30,13 +31,6 @@ class FamilyFactory
...
     protected $factory;
 
-    /**
-     * Constructor
-     *
-     * @param ProductManager              $productManager
-     * @param ChannelManager              $channelManager
-     * @param AttributeRequirementFactory $factory
-     */
     public function __construct(
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php]: where is the phpdoc ?

File: src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php [hunk 3]
```diff
@@ -68,11 +60,6 @@ public function createFamily()
         return $family;
     }
 
-    /**
-     * Get the PIM channels
-     *
-     * @return \Pim\Bundle\CatalogBundle\Entity\Channel[]
-     */
     protected function getChannels()
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php]: you removed the doc is it normal ?

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/NotBlankGuesserSpec.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        $this->supportAttribute($attribute)
+            ->shouldReturn(true);
+
+        $attribute->getAttributeType()
+            ->willReturn('foo');
+        $this->supportAttribute($attribute)
+            ->shouldReturn(true);
+    }
+
+    function it_guess_required(AttributeInterface $attribute)
```
Comment (jjanvier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/NotBlankGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UrlGuesserSpec.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    function it_does_not_support_other_attributes($identifier, $textarea, AttributeInterface $image)
+    {
+        $this->supportAttribute($identifier)->shouldReturn(false);
+        $this->supportAttribute($textarea)->shouldReturn(false);
+
+        $image->getAttributeType()->willReturn('pim_catalog_image');
+        $this->supportAttribute($image)->shouldReturn(false);
+    }
+
+    function it_guess_url($text)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UrlGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        $attribute->isUnique()->willReturn(false);
+
+        $this->guessConstraints($attribute)
+            ->shouldReturn([]);
+    }
+
+    private function dataProviderForSupportedAttributes()
+    {
+        return array(
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php]: array => []

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        foreach ($this->dataProviderForSupportedAttributes() as $attributeTypeTest) {
+            $attributeType = $attributeTypeTest[0];
+            $expectedResult = $attributeTypeTest[1];
+            $attribute->getBackendType()
+                ->willReturn($attributeType);
+            $this->supportAttribute($attribute)->shouldReturn($expectedResult);
+        }
+    }
+
+    public function it_guess_unique_value(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RegexGuesserSpec.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        $this->supportAttribute($attribute)
+            ->shouldReturn(true);
+
+        $attribute->getAttributeType()
+            ->willReturn('foo');
+        $this->supportAttribute($attribute)
+            ->shouldReturn(false);
+    }
+
+    function it_guess_regex(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RegexGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(null);
+        $constraint->max
+            ->shouldBe('2038-01-19');
+    }
+
+    function it_guess_date_min_and_max(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe('1970-01-01');
+        $constraint->max
+            ->shouldBe(null);
+    }
+
+    function it_guess_date_max(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(0);
+        $constraint->max
+            ->shouldBe(null);
+    }
+
+    function it_guess_date_min(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(-5);
+        $constraint->max
+            ->shouldBe(null);
+    }
+
+    function it_guess_non_date_min_negative_not_allowed_will_be_zero(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(5);
+        $constraint->max
+            ->shouldBe(10);
+    }
+
+    function it_guess_non_date_min_negative_allowed(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(null);
+        $constraint->max
+            ->shouldBe(10);
+    }
+
+    function it_guess_non_date_min_and_max(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+
+        $constraint = $constraints[0];
+        $constraint->shouldBeAnInstanceOf('Pim\Bundle\CatalogBundle\Validator\Constraints\Range');
+        $constraint->min
+            ->shouldBe(5);
+        $constraint->max
+            ->shouldBe(null);
+    }
+
+    function it_guess_non_date_max(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+        $this->supportAttribute($attribute)
+            ->shouldReturn(false);
+
+        $attribute->getAttributeType()
+            ->willReturn('foo');
+        $this->supportAttribute($attribute)
+            ->shouldReturn(false);
+    }
+
+    function it_guess_non_date_min(AttributeInterface $attribute)
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RangeGuesserSpec.php]: guesses

File: spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php [hunk 17]
```diff
@@ -27,25 +29,48 @@ function it_creates_a_family(
         AttributeInterface $identifierAttribute,
         ChannelInterface $printChannel,
         ChannelInterface $ecommerceChannel,
-        AttributeRequirementInterface $requirement
-    ) {
+        AttributeRequirementInterface $printRequirement,
+        AttributeRequirementInterface $ecommerceRequirement
+    )
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php]: should be one line

File: spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php [hunk 18]
```diff
@@ -16,7 +17,8 @@ function let(
         ProductManager $productManager,
         ChannelManager $channelManager,
         AttributeRequirementFactory $factory
-    ) {
+    )
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php]: should be inline

File: spec/Pim/Bundle/CatalogBundle/EventSubscriber/CreateAttributeRequirementSubscriberSpec.php [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+        $this->prePersist($eventArgs)
+            ->shouldReturn(null);
+    }
+
+    public function it_does_not_create_requirements_for_family_without_attributes(
+        $eventArgs,
+        $entityManager,
+        $repository,
+        $family
+    )
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/EventSubscriber/CreateAttributeRequirementSubscriberSpec.php]: should be inline

File: spec/Pim/Bundle/CatalogBundle/EventSubscriber/CreateAttributeRequirementSubscriberSpec.php [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+        $this->prePersist($eventArgs)
+            ->shouldReturn(null);
+    }
+
+    public function it_does_not_create_requirement_without_family(
+        $eventArgs,
+        $entityManager,
+        $repository,
+        $family
+    )
```
Comment (solivier) [inline spec/Pim/Bundle/CatalogBundle/EventSubscriber/CreateAttributeRequirementSubscriberSpec.php]: should be inline

File: src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php (hunk 3 above)
Comment (jmleroux) [inline src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php]: When PhpDoc does not give additionial information, i don't see the need to keep it.
Moreover, it has a maintenance cost.

File: src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php (hunk 2 above)
Comment (jmleroux) [inline src/Pim/Bundle/CatalogBundle/Factory/FamilyFactory.php]: I don't remember having touched this one. I'll check.

File: spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterAttributeConstraintGuessersPassSpec.php (hunk 1 above)
Comment (jmleroux) [inline spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterAttributeConstraintGuessersPassSpec.php]: :no_mouth:

File: spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php (hunk 17 above)
Comment (jmleroux) [inline spec/Pim/Bundle/CatalogBundle/Factory/FamilyFactorySpec.php]: I need to configure my IDE...

**Pull request merged**